rust/src/modbus/detect.rs

```diff
@@ -0,0 +1,563 @@
+                    '<' => Ok(std::u16::MIN..num),
+                    _ => Ok(num..num),
+                },
+                Err(_) => {
+                    SCLogError!("Invalid min number: {}", min_str);
+                    Err(())
+                }
+            }
+        } else {
+            Err(())
+        }
+    } else {
+        let min = match min_str.parse::<u16>() {
+            Ok(num) => num,
+            Err(_) => {
+                SCLogError!("Invalid min number: {}", min_str);
+                return Err(());
+            }
+        };
+
+        let max = match max_str.parse::<u16>() {
+            Ok(num) => num,
+            Err(_) => {
+                SCLogError!("Invalid max number: {}", max_str);
+                return Err(());
+            }
+        };
+
+        Ok(min..max)
+    }
+}
+
+/// Intermediary function between the C code and the access type parsing function.
+#[no_mangle]
+pub unsafe extern "C" fn rs_modbus_parse_access(c_arg: *const c_char) -> *mut c_void {
+    if c_arg.is_null() {
+        return std::ptr::null_mut();
+    }
+    if let Ok(arg) = CStr::from_ptr(c_arg).to_str() {
+        match parse_access(&arg) {
+            Ok(detect) => return Box::into_raw(Box::new(detect)) as *mut c_void,
+            Err(()) => return std::ptr::null_mut(),
+        }
+    }
+    std::ptr::null_mut()
+}
+
+/// Intermediary function between the C code and the function code parsing function.
+#[no_mangle]
+pub unsafe extern "C" fn rs_modbus_parse_function(c_arg: *const c_char) -> *mut c_void {
+    if c_arg.is_null() {
+        return std::ptr::null_mut();
+    }
+    if let Ok(arg) = CStr::from_ptr(c_arg).to_str() {
+        match parse_function(&arg) {
+            Ok(detect) => return Box::into_raw(Box::new(detect)) as *mut c_void,
+            Err(()) => return std::ptr::null_mut(),
+        }
+    }
+    std::ptr::null_mut()
+}
+
+/// Intermediary function between the C code and the unit ID parsing function.
+#[no_mangle]
+pub unsafe extern "C" fn rs_modbus_parse_unit_id(c_arg: *const c_char) -> *mut c_void {
+    if c_arg.is_null() {
+        return std::ptr::null_mut();
+    }
+    if let Ok(arg) = CStr::from_ptr(c_arg).to_str() {
+        match parse_unit_id(&arg) {
+            Ok(detect) => return Box::into_raw(Box::new(detect)) as *mut c_void,
+            Err(()) => return std::ptr::null_mut(),
+        }
+    }
+    std::ptr::null_mut()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_modbus_free(ptr: *mut c_void) {
+    if !ptr.is_null() {
+        let _ = Box::from_raw(ptr as *mut DetectModbusRust);
+    }
+}
+
+/// Compares a transaction to a signature to determine whether the transaction
+/// matches the signature. If it does, 1 is returned; otherwise 0 is returned.
+#[no_mangle]
+pub extern "C" fn rs_modbus_inspect(tx: &ModbusTransaction, modbus: &DetectModbusRust) -> u8 {
+    let msg = match &tx.request {
+        Some(r) => r,
+        None => match &tx.response {
```

Comment:
Does this mean that If a transaction has a request and a response, we only match on the request ?

Reply:
The [modbus keyword ](https://suricata.readthedocs.io/en/suricata-6.0.2/rules/modbus-keyword.html?highlight=modbus)supports unit, access and address inspection. These fields are guaranteed to be the same because of how [matches ](https://docs.rs/sawp-modbus/0.4.0/src/sawp_modbus/lib.rs.html#874-1045) associates a response to a request.

Reply:
Looking at it again there is also `value`, which may only be present in either the request or response.

Reply:
Thanks for these explanations.
Could you put comments in the code about this ?
So that future extensions to the modbus keyword do not break these expectations...